Src/Base/AMReX_VisMF.H

```diff
@@ -380,6 +365,380 @@ std::ostream& operator<< (std::ostream& os, const VisMF::Header& hd);
+        for (auto const& rv : fileNumbersWriteOrder) {
+            Long currentOffset = 0;
+            for (auto rank : rv) {
+                auto rbo_it = rankBoxOrder.find(rank);
+                if (rbo_it != rankBoxOrder.end()) {
+                    Vector<int> const& index = rbo_it->second;
+                    int whichFileNumber = fileNumbers[rank];
+                    std::string const& whichFileName =
+                        VisMF::BaseName(NFilesIter::FileName(whichFileNumber, filePrefix));
+                    for (int i : index) {
+                        fod[i].m_name = whichFileName;
+                        fod[i].m_head = currentOffset;
+                        currentOffset += fabBytes[i];
+                    }
+                }
+            }
+        }
+        ofs << fod;
+    }
+}
+
+namespace detail {
+template <typename FAB>
+void read_fab(FAB& fab, VisMF::FabOnDisk const& fod, std::string const& name)
+{
+    std::string fullname = VisMF::DirName(name);
+    fullname += fod.m_name;
+    VisMFBuffer::IO_Buffer io_buffer(VisMFBuffer::GetIOBufferSize());
+    std::ifstream ifs;
+    ifs.rdbuf()->pubsetbuf(io_buffer.dataPtr(), io_buffer.size());
+    ifs.open(fullname.c_str(), std::ios::in | std::ios::binary);
+    if (!ifs.good()) {
+        amrex::FileOpenFailed(fullname);
+    }
+    ifs.seekg(fod.m_head, std::ios::beg);
+    fab.readFrom(ifs);
+}
+}
+
+/**
+ * \brief Read iMultiFab/FabArray<IArrayBox>
+ *
+ * This reads an iMultiFab/FabArray<IArrayBox> from disk.  If it has been
+ * fully defined, the BoxArray on the disk must match the BoxArray in the
+ * given iMultiFab/FabArray<IArrayBox> object.  If it is only constructed
+ * with the default constructor, the BoxArray on the disk will be used and a
+ * new DistributionMapping will be made.  When this funciton is used to
+ * restart a calculation from checkpoint files, one should use a fully
+ * defined iMultiFab/FabArray<IArrayBox> except for the first one in a
+ * series of iMultiFab/MultiFab objects that share the same
+ * BoxArray/DistributionMapping.  This will ensure that they share the same
+ * BoxArray/DistributionMapping after restart.
+ *
+ * \param fa is the iMultiFab.
+ * \param name is the base name for the files.
+ */
+template <typename FAB>
+// This function does work for MultiFab, but let's disable it to avoid confusion.
+std::enable_if_t<std::is_same<FAB,IArrayBox>::value>
+Read (FabArray<FAB>& fa, const std::string& name)
+{
+    BL_PROFILE("Read(FabArray)");
+    AMREX_ASSERT(name.back() != '/');
+
+    BoxArray ba;
+    int ncomp;
+    IntVect ngrow;
+    Vector<VisMF::FabOnDisk> fod;
+    {
+        std::string header_file_name = name + "_H";
+        Vector<char> header_file_chars;
+        ParallelDescriptor::ReadAndBcastFile(header_file_name, header_file_chars);
+        std::string header_file_string(header_file_chars.data());
+        std::stringstream ifs(header_file_string, std::istringstream::in);
+
```

Comment:
Should we check the status of `ifs` here (and elsewhere where we read / write from streams in this PR?

Reply:
It seems that we don't usually check the state of stringstream.  I guess it was assumed that stringstream is unlikely to fail unlike fstream.  Yes, it does not hurt to check.

Reply:
In this case, there is an assertion later on the version string read from the stream.  So I guess it would fail the assertion if the stringstream is not good.